lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -0,0 +1,254 @@
+    DatabaseRequestedEvent,
+)
+from charms.opensearch.v0.constants_charm import (
+    ClientRelationBadRoleRequestMessage,
+    ClientRelationName,
+    PeerRelationName,
+)
+from charms.opensearch.v0.helper_databag import Scope
+from charms.opensearch.v0.helper_networking import units_ips
+from charms.opensearch.v0.helper_security import generate_hashed_password
+from charms.opensearch.v0.opensearch_users import OpenSearchUserMgmtError
+from ops.charm import CharmBase, RelationBrokenEvent, RelationDepartedEvent
+from ops.framework import Object
+from ops.model import BlockedStatus, Relation
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchProvider(Object):
+    """Defines functionality for the 'provides' side of the 'opensearch-client' relation.
+
+    Hook events observed:
+        - database-requested
+        - relation-departed
+        - relation-broken
+    """
+
+    def __init__(self, charm: CharmBase) -> None:
+        """Constructor for OpenSearchProvider object.
+
+        Args:
+            charm: the charm for which this relation is provided
+        """
+        super().__init__(charm, ClientRelationName)
+
+        self.charm = charm
+        self.unit = self.charm.unit
+        self.app = self.charm.app
+        self.opensearch = self.charm.opensearch
+        self.user_manager = self.charm.user_manager
+
+        self.relation_name = ClientRelationName
+        self.database_provides = DatabaseProvides(self.charm, relation_name=self.relation_name)
+
+        self.framework.observe(
+            self.database_provides.on.database_requested, self._on_database_requested
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_departed, self._on_relation_departed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_broken
+        )
+
+    def _relation_username(self, relation: Relation) -> str:
+        return f"{self.relation_name}_{relation.id}_user"
+
+    def _depart_flag(self, relation: Relation):
+        return f"{self.relation_name}_{relation.id}_departing"
+
+    def _unit_departing(self, relation):
+        return self.charm.peers_data.get(Scope.UNIT, self._depart_flag(relation))
+
+    def _on_database_requested(self, event: DatabaseRequestedEvent) -> None:
+        """Handle client database-requested event.
+
+        The read-only-endpoints field of DatabaseProvides is unused in this relation because this
+        concept is irrelevant to OpenSearch. In this relation, the application charm should have
+        control over node & index security policies, and therefore differentiating between types of
+        network endpoints is unnecessary.
+        """
+        if not self.unit.is_leader():
+            return
+        if not self.opensearch.is_node_up():
+            event.defer()
+            return
+
+        try:
+            extra_user_roles = json.loads(event.extra_user_roles)
```

Comment:
do we have ticket or PR where we rework extra_user_roles approach?

Reply:
Yep - [DPE-1276](https://warthogs.atlassian.net/browse/DPE-1276), scheduled for this week :)